Add automated testing

narendrasss opened the pull request:
Add automated end-to-end testing using [cypress](https://www.cypress.io/).

- `author.spec.js` contains all tests related to the authoring tool as laid out in the testing checklist.
- `user.spec.js` contains all tests related to user use.

Some tests from the checklist are yet to be included here due to limitations in cypress itself:

- Dragging seems to be difficult to replicate given the way our nodes are implemented so that test simply asserts that all nodes are visible to the user.

wynnset commented on templates/vue/cypress/support/roles.js:
```diff
@@ -0,0 +1,10 @@
+  },
+  subscriber: {
+    username: "narendras",
+    password: ")%AbNZgppS)uNvxT5)OdgzhX",
```
If these require set up from the user's side, please add it to the README file. Please do not use your name, but instead a more generic username

wynnset commented on templates/vue/src/components/NodeModal.vue:
```diff
@@ -104,6 +110,7 @@
               <b-form-input
                 id="node-h5p-media-duration"
                 v-model="node.mediaDuration"
+                data-testid="node-h5pDuration"
```
Can you not use the id attribute instead of the test-specific data attribute? Or the class attribute if it's repeatable? Would be much cleaner.

wynnset replied:
Hey @narendrasss still waiting for a response here.

wynnset replied:
Oops, just saw your response - ignore.

wynnset commented on templates/vue/cypress/support/utils.js:
```diff
@@ -0,0 +1,48 @@
+// TODO: Add node generator
+// TODO: Add link generator
+
+export const SITE_URL = "http://localhost:8888"
```
This might be different on other machines - please make it customizable somehow - maybe through an environment file that you would add to .gitignore? And then include instructions to add it in the README file.

narendrasss commented:
@wynnset I changed all of the constants to cypress's environment file and I added instructions to the README. For the testId, I think it's good for us to keep it this way because we immediately know that the id is used only for testing purposes (and won't be changed in the future).

wynnset commented on templates/vue/README.md:
```diff
@@ -33,3 +33,37 @@ You need to adjust your git config so the automatic linting process works correc
+    "env": {
+        "DEV_URL": "http://localhost:8888",
+        "TEST_TAPESTRY_NAME": "testing",
+        "USER_TAPESTRY_NAME": "testing-user",
```
Can we remove these? Just name them whatever - the user doesn't care about the name of the tapestries used in testing (right?)

narendrasss replied:
By the user do you mean the developers? The `TEST_TAPESTRY_NAME` doesn't matter, but the `USER_TAPESTRY_NAME` matters because it should refer to a test tapestry that was (previously) manually created.

wynnset commented on templates/vue/README.md:
```diff
@@ -33,3 +33,37 @@ You need to adjust your git config so the automatic linting process works correc
+        "ADMIN_USERNAME": "admin",
+        "ADMIN_PASSWORD": "Tapestry123!",
+        "SUBSCRIBER_USERNAME": "narendras",
+        "SUBSCRIBER_PASSWORD": ")%AbNZgppS)uNvxT5)OdgzhX"
```
Can you use admin_user / admin_pass and subscriber_user / subscriber_pass instead of the values you have set here as examples, please?

wynnset commented on templates/vue/README.md:
```diff
@@ -33,3 +33,37 @@ You need to adjust your git config so the automatic linting process works correc
+
+* `DEV_URL` is the url of your local site.
+* `TEST_TAPESTRY_NAME` is the name of a new tapestry that is automatically made (and deleted) during the authoring tests.
+* `USER_TAPESTRY_NAME` is the name of the tapestry that is used for the user-side tests.
```
As per my previous comment, let's remove these?

narendrasss replied:
Please see my other comment.

wynnset commented on templates/vue/cypress.json:
```diff
@@ -0,0 +1,11 @@
+        "SUBSCRIBER_USERNAME": "narendras",
+        "SUBSCRIBER_PASSWORD": ")%AbNZgppS)uNvxT5)OdgzhX"
+    }
+}
```
I think it's better we remove this file and add it to gitignore and just update the README to let the user know they need to add this file in this format. Otherwise, users might accidentally push their own settings to the repo, which we don't want.

narendrasss replied:
Sure, I'll add a `cypress-sample.json` and update the README
